SignalServiceKit/src/Messages/Attachments/OWSAttachmentDownloads.m

```diff
@@ -35,42 +36,104 @@
+@interface OWSAttachmentDownloadJob : NSObject
+
+@property (nonatomic, readonly) TSAttachmentPointer *attachmentPointer;
+@property (nonatomic, readonly, nullable) TSMessage *message;
+@property (nonatomic, readonly) AttachmentDownloadSuccess success;
+@property (nonatomic, readonly) AttachmentDownloadFailure failure;
 
 @end
 
-@implementation OWSAttachmentsProcessor
+#pragma mark -
+
+@implementation OWSAttachmentDownloadJob
 
 - (instancetype)initWithAttachmentPointer:(TSAttachmentPointer *)attachmentPointer
-                           networkManager:(TSNetworkManager *)networkManager
+                                  message:(nullable TSMessage *)message
+                                  success:(AttachmentDownloadSuccess)success
+                                  failure:(AttachmentDownloadFailure)failure
 {
     self = [super init];
     if (!self) {
         return self;
     }
 
-    _networkManager = networkManager;
-
-    _attachmentPointers = @[ attachmentPointer ];
-    _attachmentIds = @[ attachmentPointer.uniqueId ];
+    _attachmentPointer = attachmentPointer;
+    _message = message;
+    _success = success;
+    _failure = failure;
 
     return self;
 }
 
-- (instancetype)initWithAttachmentProtos:(NSArray<SSKProtoAttachmentPointer *> *)attachmentProtos
-                          networkManager:(TSNetworkManager *)networkManager
-                             transaction:(YapDatabaseReadWriteTransaction *)transaction
+@end
+
+#pragma mark -
+
+@interface OWSAttachmentDownloads ()
+
+// This property should only be accessed while synchronized on this class.
+@property (nonatomic, readonly) NSMutableSet<NSString *> *downloadingAttachmentIdSet;
+// This property should only be accessed while synchronized on this class.
+@property (nonatomic, readonly) NSMutableArray<OWSAttachmentDownloadJob *> *attachmentDownloadJobQueue;
+
+@end
+
+#pragma mark -
+
+@implementation OWSAttachmentDownloads
+
+#pragma mark - Dependencies
+
+- (OWSPrimaryStorage *)primaryStorage
+{
+    return SSKEnvironment.shared.primaryStorage;
+}
+
+- (TSNetworkManager *)networkManager
+{
+    return SSKEnvironment.shared.networkManager;
+}
+
+
+#pragma mark -
+
+- (instancetype)init
 {
     self = [super init];
     if (!self) {
         return self;
     }
 
-    _networkManager = networkManager;
+    _downloadingAttachmentIdSet = [NSMutableSet new];
+    _attachmentDownloadJobQueue = [NSMutableArray new];
+
+    return self;
+}
+
+#pragma mark -
+
+- (NSArray<NSString *> *)attachmentsIdsForAttachments:(NSArray<TSAttachment *> *)attachments
+{
+    OWSAssertDebug(attachments);
 
     NSMutableArray<NSString *> *attachmentIds = [NSMutableArray new];
+    for (TSAttachment *attachment in attachments) {
+        [attachmentIds addObject:attachment.uniqueId];
+    }
+    return [attachmentIds copy];
+}
+
+- (NSArray<TSAttachmentPointer *> *)
+    saveAttachmentPointersForAttachmentProtos:(NSArray<SSKProtoAttachmentPointer *> *)attachmentProtos
```

Comment:
As we discussed this AM (I think), it makes sense for most of the logic in this method to live on some kind of `TSAttachmentPointer#fromProto` method.

Reply:
Yup!